src/metabase/automagic_dashboards/core.clj

```diff
@@ -653,109 +444,6 @@
                              :score         score))))))))
 
-(defn- matching-rules
-  "Return matching rules orderd by specificity.
-   Most specific is defined as entity type specification the longest ancestor
-   chain."
-  [rules {:keys [source entity]}]
-  (let [table-type (or (:entity_type source) :entity/GenericTable)]
-    (->> rules
-         (filter (fn [{:keys [applies_to]}]
-                   (let [[entity-type field-type] applies_to]
-                     (and (isa? table-type entity-type)
-                          (or (nil? field-type)
-                              (field-isa? entity field-type))))))
-         (sort-by :specificity >))))
-
-(defn- linked-tables
-  "Return all tables accessable from a given table with the paths to get there.
-   If there are multiple FKs pointing to the same table, multiple entries will
-   be returned."
-  [table]
-  (for [{:keys [id target]} (field/with-targets
-                              (db/select Field
-                                :table_id           (u/get-id table)
-                                :fk_target_field_id [:not= nil]
-                                :active             true))
-        :when (some-> target mi/can-read?)]
-    (-> target field/table (assoc :link id))))
-
-(def ^:private ^{:arglists '([source])} source->engine
-  (comp :engine Database (some-fn :db_id :database_id)))
-
-(defmulti
-  ^{:private  true
-    :arglists '([context entity])}
-  inject-root (fn [_ entity] (type entity)))
-
-(defmethod inject-root (type Field)
-  [context field]
-  (let [field (assoc field
-                :link   (->> context
-                             :tables
-                             (m/find-first (comp #{(:table_id field)} u/get-id))
-                             :link)
-                :engine (-> context :source source->engine))]
-    (update context :dimensions
-            (fn [dimensions]
-              (->> dimensions
-                   (keep (fn [[identifier definition]]
-                           (when-let [matches (->> definition
-                                                   :matches
-                                                   (remove (comp #{(id-or-name field)} id-or-name))
-                                                   not-empty)]
-                             [identifier (assoc definition :matches matches)])))
-                   (concat [["this" {:matches [field]
-                                     :name    (:display_name field)
-                                     :score   rules/max-score}]])
-                   (into {}))))))
-
-(defmethod inject-root (type Metric)
-  [context metric]
-  (update context :metrics assoc "this" {:metric (->reference :mbql metric)
-                                         :name   (:name metric)
-                                         :score  rules/max-score}))
-
-(defmethod inject-root :default
-  [context _]
-  context)
-
-(s/defn ^:private make-context
```

Comment:
did all this go to the MBTL codebase or is it going to be re-implemented elsewhere?

Reply:
Some went to MBTL, the rest is no longer needed. A lot of complexity was there to support varied shapes of data (via overloading and various ways of binding), which we no longer need to do as x-rays will just query the data model, and the data model decomposes the problem to: start from something known (& fixed), and than transform.

Reply:
that's a good sign!